### PR TITLE
ADAU1701 support

### PR DESCRIPTION
**Pull request by m0n5t3r**
This is the second leg of #12; it includes

* some tweaks to the DSP base class (some additional common bits moved to it, some abstract methods added)
* the actual ADAU1701 chip handler
* SigmaStudio protocol handler common parts went into a base class, the existing protocol implementation (TCPIOADAU145x) went into its own module and a new one for TCPIP1701 got added

it depends on changes in #13, so it will need to be merged after that one; I'll deal with whatever conflicts might arise if you do a squashed merge for #13 when the time comes

### CODE REVIEW

**Comment by m0n5t3r** [comment]
right now it shows the whole diff, of course; it should become more manageable after the merge of #13

**Comment by elagil** [comment]
Thanks for your work! I didn't have time to review yet, but I will try to do it as soon as possible.

**Comment by elagil** [comment]
If you don't mind, I will wait with the review until we merge https://github.com/elagil/sigmadsp/pull/13. This is a bit hard to distinguish otherwise.

> I'll deal with whatever conflicts might arise if you do a squashed merge for https://github.com/elagil/sigmadsp/pull/13 when the time comes

Why would there be conflicts if I squashed the commits? The files would not change.

File: sigmadsp/application/backend.py
Comment (elagil) [inline sigmadsp/application/backend.py:162]:
```suggestion
            elif isinstance(request, SafeloadRequest):
```

File: sigmadsp/application/backend.py
Comment (elagil) [inline sigmadsp/application/backend.py:167]: We may add an `else` case that raises an error, e.g. `raise TypeError(f"Unknown command type {type(request)}.")`.

File: sigmadsp/communication/base.py
Comment (elagil) [inline sigmadsp/communication/base.py:16]: You could just inherit, if it's the same as `WriteRequest` anyway.

File: sigmadsp/communication/sigma_tcp_server.py
Comment (elagil) [inline sigmadsp/communication/sigma_tcp_server.py:27]: I don't think the DSP type should be optional.

File: sigmadsp/communication/sigma_tcp_server.py
Comment (elagil) [inline sigmadsp/communication/sigma_tcp_server.py:54]:
```suggestion
        protocol_handler: 
        
        if self.dsp_type == "adau145x":
            protocol_handler = tcpipadau145x.SigmaStudioRequestHandler

        elif self.dsp_type == "adau1701":
            protocol_handler = tcpip1701.SigmaStudioRequestHandler
        
        else:
            raise TypeError(f"The specified DSP type {self.dsp_type} is not supported.")
```

File: sigmadsp/communication/tcpip1701.py
Comment (elagil) [inline sigmadsp/communication/tcpip1701.py:27]: This is borrowing a lot from the old ADAU145x implementation. Most functionality is identical. I suggest to make a shared `SigmaStudioRequestHandler` base class and inherit from it.

File: sigmadsp/communication/tcpip1701.py
Comment (elagil) [inline sigmadsp/communication/tcpip1701.py:1]: this module's file name does not match the scheme you chose for `tcpipadau145x.py`

I would prefer `tcpip_adau145x.py` and `tcpip_adau1701.py` or even `requests_handling_adau145x.py` and `requests_handling_adau1701.py`.

File: sigmadsp/hardware/adau1701.py
Comment (elagil) [inline sigmadsp/hardware/adau1701.py:52]: It would be nice to inform the frontend of this, because logger messages are not visible there. Otherwise, you could fall back to `hard_reset`.

File: sigmadsp/helper/conversion.py
Comment (elagil) [inline sigmadsp/helper/conversion.py:55]: Seeing this, there should probably be an overflow check. Also, what happens for negative `value`s? Of course, these problems also affect the old implementation.

**Review by elagil — CHANGES_REQUESTED** [review]
Some initial review comments

File: sigmadsp/communication/tcpip1701.py
Comment (m0n5t3r) [reply sigmadsp/communication/tcpip1701.py:1]: I used the names of the protocol blocks in sigma studio; also, something like `requests_handling_adau145x` is pointlessly verbose IMO

File: sigmadsp/hardware/adau1701.py
Comment (m0n5t3r) [reply sigmadsp/hardware/adau1701.py:52]: https://github.com/elagil/sigmadsp/blob/main/sigmadsp/hardware/dsp.py#L189-L192

File: sigmadsp/helper/conversion.py
Comment (m0n5t3r) [reply sigmadsp/helper/conversion.py:55]: the 8.24 / 5.23 values are signed (-128 to 128 / -16 to 16); not sure about overflow checks, I'd rather clamp the input to avoid surprises (wrap-around)

File: sigmadsp/communication/tcpip1701.py
Comment (m0n5t3r) [reply sigmadsp/communication/tcpip1701.py:27]: it already has a lot of the common parts moved into the parent class, I might have missed a few and maybe other things can be parameterized to deal with the differences; I'll take another look when I have time

File: sigmadsp/communication/sigma_tcp_server.py
Comment (m0n5t3r) [reply sigmadsp/communication/sigma_tcp_server.py:27]: same as before, I went for "change things without breaking things"; since we already require changes to the config, I'll make this mandatory

File: sigmadsp/communication/tcpip1701.py
Comment (m0n5t3r) [reply sigmadsp/communication/tcpip1701.py:1]: ah, I remembered wrong, nevermind; will gather the common parts in a parent class

File: sigmadsp/hardware/adau1701.py
Comment (elagil) [reply sigmadsp/hardware/adau1701.py:52]: True, I forgot. This would make a nice infinite loop.

File: sigmadsp/communication/tcpip1701.py
Comment (m0n5t3r) [reply sigmadsp/communication/tcpip1701.py]: I ended up doing a bit more...

File: sigmadsp/communication/sigmastudio_protocols.py
Comment (elagil) [inline sigmadsp/communication/sigmastudio_protocols.py:16]: I have a rather lengthy suggestion, which would make the code easier to follow (in my opinion), and allows for error checks. I would introduce classes for a `Field` and the collection of fields, which is `HeaderFields`.

Let me know what you think:

```suggestion
@dataclass
class Field:
    # The name of the field, e.g. "operation", "safeload", "channel".
    name: str

    # The offset of the field in bytes from the start of the header.
    offset: int

    # The size of the field in bytes.
    size: int

    # The stored value.
    value: int = 0

    def __post_init__(self):
        if self.size < 0:
            raise ValueError("Field size must be a positive integer.")

        if self.offset < 0:
            raise ValueError("Field offset must be a positive integer.")

        self.end = self.offset + self.size - 1


class HeaderFields:
    def __init__(self):
        self._fields: Set[Field] = set()

    @property
    def fields(self) -> Set[Field]:
        """The fields of the header, sorted by their offset."""
        return sorted(self._fields, key=lambda field: field.offset)

    @property
    def length(self) -> int:
        """The total length of the header."""
        return sum([field.size for field in self.fields])

    @property
    def continuous(self) -> bool:
        """Whether or not there are spaces in the header that are not defined."""
        for field, next_field in zip(self.fields, self.fields[1:]):
            if (field.end + 1) != next_field.offset:
                return False

        return True

    def _sanity_check(self):
        """Check for overlapping fields.

        Raises:
            MemoryError: If overlapping fields are found.
        """
        # Check for overlapping fields, which are sorted by their offset.
        for field, next_field in zip(self.fields, self.fields[1:]):
            if field.end <= next_field.offset:
                raise MemoryError("Fields {field.name} and {next_field.name} overlap.")

    def add(self, field: Field):
        """Add a new field. Duplicates are ignored.

        Args:
            field (Field): The field to add.
        """
        self._fields.append(field)

    def get_field_by_name(self, name: str) -> Union[Field, None]:
        """Get a field by its name.

        Args:
            name (str): The name of the field.

        Returns:
            Union[Field, None]: The field, or None, if no field was found.
        """
        for field in self._fields:
            if field.name == name:
                return field

        return None
```

**Review by elagil — COMMENTED** [review]
A suggestion from my side. I will review the rest tomorrow.

File: sigmadsp/communication/sigmastudio_protocols.py
Comment (m0n5t3r) [reply sigmadsp/communication/sigmastudio_protocols.py:16]: it's overly complicated, how do you even use that... anyway, I'll apply it if you give me a hint on how it's used; I just want to get this merged and then never touch this style of Python again...

Comment (elagil) [reply sigmadsp/communication/sigmastudio_protocols.py:15]: As I say, it is only a suggestion and you don't have to use it. I just find it easier to read than dictionaries, where the fields have no names.

Comment (m0n5t3r) [reply sigmadsp/communication/sigmastudio_protocols.py:15]: Well, you'll be stuck with it long term, if your mind works better with those dataclasses and class hierarchies when dicts are easy to read at these dimensions and give you ordering for free. I find this style exhausting to read (with the configuration parser it took me like half an hour to figure out how I should name my volume control, and I still got it wrong because it seemed to me `target` had to be in the name), and half the time I spent coding was spent figuring out clever ways to avoid import loops and making a clean language look like a very ugly version of Go...

Can you make the suggestion checks-compliant and add an usage example in the `HeaderFields` class docstring? I'm not being passive aggressive, I genuinely have no idea how I'd use this (and what I'd add to the docstrings because at least one of those annoying checks will complain).

Comment (elagil) [reply sigmadsp/communication/sigmastudio_protocols.py:15]: > with the configuration parser it took me like half an hour to figure out how I should name my volume control, and I still got it wrong because it seemed to me `target` had to be in the name

Well, I am not saying it's perfect. The variable naming can certainly be improved in many areas.


> Can you make the suggestion checks-compliant and add an usage example in the `HeaderFields` class docstring?

I don't want to force you to adopt my `dataclass`-heavy style, so I created a pull request on top of yours. See https://github.com/m0n5t3r/sigmadsp/pull/1. The usage should be very similar to the previous dicts, but adds some comfort methods.

Comment (m0n5t3r) [reply sigmadsp/communication/sigmastudio_protocols.py:17]: merged, confirmed it works; I don't think OrderedDict was strictly necessary (turns out dicts are ordered in Python 3.6 and later), but I guess it helps compatibility with older versions.

**Review by elagil — APPROVED** [review]
Thanks again for your work! I have no further complaints. I really like the abstraction with the `_REGISTRY` that handles protocols across chipsets.

I will merge after testing on my hardware.

File: sigmadsp/communication/sigma_tcp_server.py
Comment (elagil) [inline sigmadsp/communication/sigma_tcp_server.py:122]: This is incompatible with the backend's check for the dsp type, which only accepts the string `adau14xx`. I find it dangerous to compare against dsp_type strings which are defined in different places.

As a fix, this should read:
```suggestion
    dsp_type = "adau14xx"
```

File: sigmadsp/communication/sigma_tcp_server.py
Comment (elagil) [inline sigmadsp/communication/sigma_tcp_server.py:119]:
```suggestion
class Adau14xxRequestHandler(SigmaStudioRequestHandler):
```

File: sigmadsp/communication/sigma_tcp_server.py
Comment (elagil) [inline sigmadsp/communication/sigma_tcp_server.py:162]:
```suggestion
        if self.dsp_type == "adau14xx":
```

**Review by elagil — CHANGES_REQUESTED** [review]
I found a bug while testing: The application did not accept my old dsp setting of `adau14xx`, but also not `adau145x`. This is because the type checks in `backend` and the `sigma_tcp_server` do not use the same basis for verifying compatibility.

In a refactoring PR, I will try to address this by having a single point of "truth" for supported chips. For now, let's just make it equal across modules.

File: sigmadsp/communication/sigma_tcp_server.py
Comment (m0n5t3r) [reply sigmadsp/communication/sigma_tcp_server.py:122]: it will be misleading, tho - 14xx includes 1401, which is basically automotive-grade 1701

Comment (m0n5t3r) [reply sigmadsp/communication/sigma_tcp_server.py:122]: ah, I see now, in `sigmadsp.hardware` it's the old version / 14xx; I'll commit these and leave it to the single-source-of-truth changes to address the misleading part

**Comment by elagil** [comment]
Looks good! 😃

**Pull request merged**
